python/ray/serve/async_goal_manager.py

```diff
@@ -8,34 +8,48 @@
 
 
 class AsyncGoalManager:
+    """
+    Class to help ServeController to manage desirable states of replica.
+
+    Essentially a dictionary of <goal_id, asyncio.Future> within
+    ServeController's event loop with APIs to facilitate common calls.
+    """
+
     def __init__(self):
-        self._pending_goals: Dict[GoalId, asyncio.Event] = dict()
+        self._pending_goals: Dict[GoalId, asyncio.Future] = dict()
 
     def num_pending_goals(self) -> int:
         return len(self._pending_goals)
 
     def create_goal(self, goal_id: Optional[GoalId] = None) -> GoalId:
         if goal_id is None:
             goal_id = uuid4()
-        self._pending_goals[goal_id] = asyncio.Event()
+
+        self._pending_goals[
+            goal_id] = asyncio.get_running_loop().create_future()
         return goal_id
 
     def complete_goal(self, goal_id: GoalId) -> None:
         logger.debug(f"Completing goal {goal_id}")
-        event = self._pending_goals.pop(goal_id, None)
-        if event:
-            event.set()
+        future = self._pending_goals.pop(goal_id, None)
+        if future:
+            future.set_result(goal_id)
 
     def check_complete(self, goal_id: GoalId) -> bool:
+        # TODO: Maybe we can switch this to fut.done() instead ?
         return goal_id not in self._pending_goals
 
     async def wait_for_goal(self, goal_id: GoalId) -> None:
         start = time.time()
         if goal_id not in self._pending_goals:
             logger.debug(f"Goal {goal_id} not found")
             return
+        try:
+            await self._pending_goals[goal_id]
+        except Exception as e:
+            # Exception thrown won't stop controller main loop, only
+            # handled by goal_id's final status
+            self._pending_goals[goal_id].set_exception(e)
```

Comment:
It be safer to do `self._pending_goals.get(goal_id, asyncio.Future()).set_exception(e)` to avoid some situation where this `Except` block is called when the future is removed early?

Reply:
I see what you're suggesting .. but I wonder if that adds value to create a new Future instance out of event loop's API since no one should be waiting for its return value. Following pattern from current implementation we can get the actual future obj before try catch, thus we're always awaiting and setting exception on the same obj regardless if it's already poped from pending goals.

Reply:
I think what we want to do is get a reference to the future outside the try-except then use it instead of doing another lookup in the dictionary:
```python
fut = self._pending_goals[goal_id]
try:
    await fut
except Exception as e:
    fut.set_exception(e)
```

Reply:
i don't think this is going to work? because `await self._pendings[goal_id]` means the future is already completed and has a result. you can't set_exception on a completed future. 
this block should just be `await self._pending_goals`

Reply:
I think the actual behavior for this file is not surfaced yet due to this commit directly called delete_deployment, but i totally agree with comments here. Let me change this a bit with prints as test plan on this PR so we can see a complete flow of what happened.

Just to confirm, i'm proposing the ideal state for async goals in case of constructor failure should be:
 - We throw exception from backend_worker's async function, which goes through backend_state_manager first;
 - But then we should let the exception surface to controller's async goal manager eventually, thus able to set the future tied up to original goal_id, therefore be able to surface it back to user script that called deploy() since the goal_id the represents blocking await for deploying returned with exception

Does this make sense as our deploy() API error handling ?